### PR TITLE
ViT: Load Jax augreg weights

### LINKED ISSUE #2453: Add ability to load official ViT JAX weights directly from Google Bucket
## 🚀 Feature

Currently there is `kornia.contrib.vit` but there are no pretrained weights for them. There are many official JAX checkpoints from Google, and the author also recommends some checkpoints for each model variant.

https://github.com/google-research/vision_transformer

Having the ability to directly load those weights would be convenient. Since the weights are in `.npy` format, we only need numpy to read it. Numpy can be an optional dependency (which means kornia will be able to load JAX weights only if numpy is installed).

Example

```python
from kornia.contrib.vit import VisionTransformer

url = "https://storage.googleapis.com/vit_models/augreg/B_16-i21k-300ep-lr_0.001-aug_medium1-wd_0.1-do_0.0-sd_0.0.npz"

model = VisionTransformer()  # default is ViT-B/16
model.from_jax_weights(url)

# we may also have
from kornia.contrib.vit import vit_b_16

model = vit_b_16(pretrained=True)
```

## Additional context

timm also has a similar function

https://github.com/huggingface/pytorch-image-models/blob/394e8145551191ae60f672556936314a20232a35/timm/models/vision_transformer.py#L723-L727

### PR DESCRIPTION
**Pull request by gau-nernst**
#### Changes

Fixes #2453 

Some architecture code changes
- Change hidden dim in FFN from embed_dim to embed_dim * 4
- Change LayerNorm eps from 1e-5 (PyTorch's default) to 1e-6 (Jax's default)
  - This change is necessary to get Jax weights work correctly
  - Might make this an `__init__` argument? i.e. `norm_eps`
- Add a LayerNorm layer at the end

I haven't added a test to check if the loaded weights are correct. Currently I use the code snippet below to check for correctness.

```python
from kornia.contrib.vit import VisionTransformer
import torch
import timm

model = VisionTransformer.from_config("vit_ti/16", pretrained=True).eval()
model_ref = timm.create_model("vit_tiny_patch16_224.augreg_in21k", pretrained=True).eval()

inputs = torch.randn(1, 3, 224, 224)
with torch.no_grad():
    outputs = model(inputs)
    outputs_ref = model_ref.forward_features(inputs)

diff = outputs - outputs_ref
print(diff.abs().mean())
torch.testing.assert_close(outputs, outputs_ref)
```

On my Mac M1, the average absolute error is `9.1483e-06`. `torch.testing.assert_close` returns

```
Mismatched elements: 9049 / 37824 (23.9%)
Greatest absolute difference: 8.487701416015625e-05 at index (0, 178, 115) (up to 1e-05 allowed)
Greatest relative difference: 0.025102879852056503 at index (0, 115, 17) (up to 1.3e-06 allowed)
```

The difference is due to `F.scaled_dot_product_attention()` (which timm uses). If I change the attention implementation to `F.scaled_dot_product_attention()`, the outputs are identical.

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/contrib/vit.py [hunk 1]
```diff
@@ -233,4 +261,69 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
 
         out = self.patch_embedding(x)
         out = self.encoder(out)
+        out = self.norm(out)
         return out
+
+    @torch.no_grad()
+    def load_jax_checkpoint(self, checkpoint: str) -> VisionTransformer:
+        import numpy as np
```
Comment (johnnv1) [inline kornia/contrib/vit.py:269]: cc @edgarriba what do you think would be the best approach here?

File: kornia/contrib/vit.py [hunk 2]
```diff
@@ -233,4 +261,69 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
 
         out = self.patch_embedding(x)
         out = self.encoder(out)
+        out = self.norm(out)
         return out
+
+    @torch.no_grad()
+    def load_jax_checkpoint(self, checkpoint: str) -> VisionTransformer:
```
Comment (johnnv1) [inline kornia/contrib/vit.py:268]: needs docstrings

File: kornia/contrib/vit.py [hunk 3]
```diff
@@ -233,4 +261,69 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            block[1].fn[1][0].bias.copy_(_get(prefix + "MlpBlock_3/Dense_0/bias"))
+            block[1].fn[1][3].weight.copy_(_get(prefix + "MlpBlock_3/Dense_1/kernel").T)
+            block[1].fn[1][3].bias.copy_(_get(prefix + "MlpBlock_3/Dense_1/bias"))
+
+        self.norm.weight.copy_(_get("Transformer/encoder_norm/scale"))
+        self.norm.bias.copy_(_get("Transformer/encoder_norm/bias"))
+        return self
+
+    @staticmethod
+    def from_config(variant: str, pretrained: bool = False, **kwargs: Any) -> VisionTransformer:
```
Comment (johnnv1) [inline kornia/contrib/vit.py:310]: needs docs

File: tests/contrib/test_contrib.py [hunk 4]
```diff
@@ -76,6 +76,11 @@ def backbone_mock(x):
         out = vit(img)
         assert out.shape == (1, 197, 128)
 
+    @pytest.mark.parametrize("variant", ["vit_ti/16", "vit_b/16"])
+    def test_from_config(self, variant):
+        model = kornia.contrib.VisionTransformer.from_config(variant)
+        assert isinstance(model, kornia.contrib.VisionTransformer)
```
Comment (johnnv1) [inline tests/contrib/test_contrib.py:82]:
```suggestion
    def test_from_config(self, variant):
        model = kornia.contrib.VisionTransformer.from_config(variant)
        assert isinstance(model, kornia.contrib.VisionTransformer)
    def test_from_config_pretrained(self):
        model = kornia.contrib.VisionTransformer.from_config('vit_ti/16', pretrained=True)
        assert isinstance(model, kornia.contrib.VisionTransformer)
```

File: tests/contrib/test_contrib.py [hunk 5]
```diff
@@ -76,6 +76,11 @@ def backbone_mock(x):
...
+
```
Comment (johnnv1) [inline tests/contrib/test_contrib.py:83]:
```suggestion
    def test_from_config_pretrained(self):
        model = kornia.contrib.VisionTransformer.from_config('vit_ti/16', pretrained=True)
        assert isinstance(model, kornia.contrib.VisionTransformer)
```

File: tests/contrib/test_contrib.py [hunk 6]
```diff
@@ -76,6 +76,11 @@ def backbone_mock(x):
         out = vit(img)
         assert out.shape == (1, 197, 128)
 
+    @pytest.mark.parametrize("variant", ["vit_ti/16", "vit_b/16"])
+    def test_from_config(self, variant):
```
Comment (johnnv1) [inline tests/contrib/test_contrib.py:80]:
```suggestion
    @pytest.mark.parametrize("variant", ["vit_ti/16", "vit_b/16", ""vit_l/16"])
    def test_from_config(self, variant):
```

File: kornia/contrib/vit.py [hunk 7]
```diff
@@ -6,22 +6,49 @@
...
+    "vit_l/16": "L_16-i21k-300ep-lr_0.001-aug_strong1-wd_0.1-do_0.0-sd_0.0.npz",
+    "vit_b/16": "B_16-i21k-300ep-lr_0.001-aug_medium1-wd_0.1-do_0.0-sd_0.0.npz",
+    "vit_s/16": "S_16-i21k-300ep-lr_0.001-aug_light1-wd_0.03-do_0.0-sd_0.0.npz",
+    "vit_ti/16": "Ti_16-i21k-300ep-lr_0.001-aug_none-wd_0.03-do_0.0-sd_0.0.npz",
+    "vit_b/32": "B_32-i21k-300ep-lr_0.001-aug_light1-wd_0.1-do_0.0-sd_0.0.npz",
+    "vit_s/32": "S_32-i21k-300ep-lr_0.001-aug_none-wd_0.1-do_0.0-sd_0.0.npz",
+}
+
+
+def download_to_torch_hub(url: str, progress: bool = True) -> str:
```
Comment (johnnv1) [inline kornia/contrib/vit.py:35]:
```suggestion
def _download_to_torch_hub(url: str, progress: bool = True) -> str:
```

File: tests/contrib/test_contrib.py (hunk 5 above)
Comment (johnnv1) [reply tests/contrib/test_contrib.py:83]: just to cover the code you added

File: kornia/contrib/vit.py (hunk 1 above)
Comment (gau-nernst) [reply kornia/contrib/vit.py:269]: Another approach is to have a function that converts numpy checkpoint to PyTorch's state dict, then the model can just `.load_state_dict()`

Comment (johnnv1) [reply kornia/contrib/vit.py:269]: well, If you think it makes sense to make the effort, we can convert these weights and upload them in[ hf/kornia](https://huggingface.co/kornia)

Comment (gau-nernst) [reply kornia/contrib/vit.py:269]: Sure, let's do that.

Comment (gau-nernst) [reply kornia/contrib/vit.py:269]:
```python
import subprocess
import os

import numpy as np
import torch

from kornia.contrib import VisionTransformer


def convert_jax_checkpoint(np_state_dict: dict[str, np.ndarray]):
    def get_weight(key: str) -> torch.Tensor:
        return torch.from_numpy(np_state_dict[key])

    state_dict = dict()
    state_dict["patch_embedding.cls_token"] = get_weight("cls")
    state_dict["patch_embedding.backbone.weight"] = get_weight("embedding/kernel").permute(3, 2, 0, 1)  # conv weight
    state_dict["patch_embedding.backbone.bias"] = get_weight("embedding/bias")
    state_dict["patch_embedding.positions"] = get_weight("Transformer/posembed_input/pos_embedding").squeeze(0)

    # for i, block in enumerate(self.encoder.blocks):
    for i in range(100):
        prefix1 = f"encoder.blocks.{i}"
        prefix2 = f"Transformer/encoderblock_{i}"

        if f"{prefix2}/LayerNorm_0/scale" not in np_state_dict:
            break

        state_dict[f"{prefix1}.0.fn.0.weight"] = get_weight(f"{prefix2}/LayerNorm_0/scale")
        state_dict[f"{prefix1}.0.fn.0.bias"] = get_weight(f"{prefix2}/LayerNorm_0/bias")

        mha_prefix = f"{prefix2}/MultiHeadDotProductAttention_1"
        qkv_weight = [get_weight(f"{mha_prefix}/{x}/kernel") for x in ["query", "key", "value"]]
        qkv_bias = [get_weight(f"{mha_prefix}/{x}/bias") for x in ["query", "key", "value"]]
        state_dict[f"{prefix1}.0.fn.1.qkv.weight"] = torch.cat(qkv_weight, 1).flatten(1).T
        state_dict[f"{prefix1}.0.fn.1.qkv.bias"] = torch.cat(qkv_bias, 0).flatten()
        state_dict[f"{prefix1}.0.fn.1.projection.weight"] = get_weight(f"{mha_prefix}/out/kernel").flatten(0, 1).T
        state_dict[f"{prefix1}.0.fn.1.projection.bias"] = get_weight(f"{mha_prefix}/out/bias")

        state_dict[f"{prefix1}.1.fn.0.weight"] = get_weight(f"{prefix2}/LayerNorm_2/scale")
        state_dict[f"{prefix1}.1.fn.0.bias"] = get_weight(f"{prefix2}/LayerNorm_2/bias")
        state_dict[f"{prefix1}.1.fn.1.0.weight"] = get_weight(f"{prefix2}/MlpBlock_3/Dense_0/kernel").T
        state_dict[f"{prefix1}.1.fn.1.0.bias"] = get_weight(f"{prefix2}/MlpBlock_3/Dense_0/bias")
        state_dict[f"{prefix1}.1.fn.1.3.weight"] = get_weight(f"{prefix2}/MlpBlock_3/Dense_1/kernel").T
        state_dict[f"{prefix1}.1.fn.1.3.bias"] = get_weight(f"{prefix2}/MlpBlock_3/Dense_1/bias")

    state_dict["norm.weight"] = get_weight("Transformer/encoder_norm/scale")
    state_dict["norm.bias"] = get_weight("Transformer/encoder_norm/bias")
    return state_dict


# recommended checkpoint from https://github.com/google-research/vision_transformer
base_url = "https://storage.googleapis.com/vit_models/augreg"
checkpoint_dict = {
    "vit_l/16": "L_16-i21k-300ep-lr_0.001-aug_strong1-wd_0.1-do_0.0-sd_0.0.npz",
    "vit_b/16": "B_16-i21k-300ep-lr_0.001-aug_medium1-wd_0.1-do_0.0-sd_0.0.npz",
    "vit_s/16": "S_16-i21k-300ep-lr_0.001-aug_light1-wd_0.03-do_0.0-sd_0.0.npz",
    "vit_ti/16": "Ti_16-i21k-300ep-lr_0.001-aug_none-wd_0.03-do_0.0-sd_0.0.npz",
    "vit_b/32": "B_32-i21k-300ep-lr_0.001-aug_light1-wd_0.1-do_0.0-sd_0.0.npz",
    "vit_s/32": "S_32-i21k-300ep-lr_0.001-aug_none-wd_0.1-do_0.0-sd_0.0.npz",
}
variant = "vit_ti/16"

if not os.path.exists(checkpoint_dict[variant]):
    subprocess.run(["wget", f"{base_url}/{checkpoint_dict[variant]}"], check=True)

np_state_dict = np.load(checkpoint_dict[variant])
state_dict = convert_jax_checkpoint(np_state_dict)

# confirm the model can load the weights
vit_model = VisionTransformer.from_config(variant)
vit_model.load_state_dict(state_dict)
```

This is the code to convert to PyTorch's state dict. Change `variant` accordingly to download and convert other variants. You can run it and save the weights, and upload to HF/kornia. Once you uploaded the weights, I will update the code to load the weights from there.

File: kornia/contrib/vit.py [hunk 8]
```diff
@@ -6,22 +6,49 @@
...
+    torch_hub_dir = torch.hub.get_dir()
+    filename = os.path.basename(url)
```
Comment (edgarriba) [inline kornia/contrib/vit.py:37]: Use Path from pathlib

File: kornia/contrib/vit.py (hunk 1 above)
Comment (edgarriba) [reply kornia/contrib/vit.py:269]: Sound good

Comment (johnnv1) [reply kornia/contrib/vit.py:269]: we can rename them if you have better names in mind

- vit_l/16 - https://huggingface.co/kornia/vit_l16_i21k  -- 1210Mb
- vit_b/16 - https://huggingface.co/kornia/vit_b16_i21k -- 343 Mb
- vit_s/16 - https://huggingface.co/kornia/vit_s16_i21k -- 86.7Mb
- vit_ti/16 - https://huggingface.co/kornia/vit_ti16_i21k -- 22.1Mb
- vit_b/32 - https://huggingface.co/kornia/vit_b32_i21k -- 350Mb
- vit_s/32 - https://huggingface.co/kornia/vit_s32_i21k -- 90Mb

Comment (gau-nernst) [reply kornia/contrib/vit.py:269]: @johnnv1 These weights actually come from [AugReg](https://arxiv.org/abs/2106.10270), not the original ViT paper, and the [Google's repo](https://github.com/google-research/vision_transformer#available-vit-models) recommends to use these AugReg weights instead. Perhaps we can add the word "augreg" somewhere in the name? (for reference: https://huggingface.co/timm/vit_tiny_patch16_224.augreg_in21k)

Comment (gau-nernst) [reply kornia/contrib/vit.py:269]: Also, should we also include the image size in the name? i.e. 224. Similar to https://huggingface.co/kornia/efficientvit_imagenet_b1_r224 and timm models

Comment (johnnv1) [reply kornia/contrib/vit.py]: Update the suffix to `..._augreg_i21k_r384` let me know your thoughts (@gau-nernst @edgarriba )...

The model card I believe you can send updates to it as PR on the hugging faces hub later... i just added a minimal info

Comment (gau-nernst) [reply kornia/contrib/vit.py]: Awesome! Just a small edit. The input size should be 224, not 384. For these AugReg weights, the ImageNet-21k pre-training is 224x224 (which we used). The fine-tuned version on ImageNet-1k is 384x384.

Comment (johnnv1) [reply kornia/contrib/vit.py]: ouch, will rename it again

Comment (johnnv1) [reply kornia/contrib/vit.py]: https://huggingface.co/kornia/vit_s16_augreg_i21k_r224

File: kornia/contrib/vit.py [hunk 9]
```diff
@@ -233,4 +237,53 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        out = self.norm(out)
         return out
+
+    @staticmethod
+    def from_config(variant: str, pretrained: bool = False, **kwargs: Any) -> VisionTransformer:
+        """Build ViT model based on the given config string. The format is `vit_{size}/{patch_size}`.
+        E.g. `vit_b/16` means ViT-Base, patch size 16x16. If `pretrained=True`, AugReg weights are loaded.
+        The weights are hosted on HuggingFace's model hub: https://huggingface.co/kornia.
+
+        Args:
```
Comment (johnnv1) [inline kornia/contrib/vit.py:249]: add something like this, for users now what variant's are available, and each ones have weights for it 

```suggestion
        The weights are hosted on HuggingFace's model hub: https://huggingface.co/kornia.

         .. note:: The available weights are: ...

        Args:
```

**Review by johnnv1 — APPROVED** [review]
